File: backend/src/modules/producto/admin-producto.controller.ts
import {
  Controller,
  Post,
  Get,
  Put,
  Patch,
  Delete,
  Param,
  Body,
  UploadedFiles,
  UseGuards,
  UseInterceptors,
  BadRequestException,
  HttpCode,
  HttpStatus,
} from "@nestjs/common";
import { FilesInterceptor } from "@nestjs/platform-express";
import {
  ApiTags,
  ApiOperation,
  ApiResponse,
  ApiBearerAuth,
  ApiConsumes,
  ApiBody,
} from "@nestjs/swagger";
import { memoryStorage } from "multer";
import { JwtAuthGuard } from "@modules/auth/guards/jwt-auth.guard";
import { ProductoService } from "./producto.service";
import { ImagenService } from "./imagen.service";
import { PublicarProductoDto } from "./dto/publicar-producto.dto";
import { ActualizarProductoDto } from "./dto/actualizar-producto.dto";
import { CreateProductoDto } from "./dto/create-producto.dto";
import { UpdateProductoDto } from "./dto/update-producto.dto";

@ApiTags("admin/productos")
@ApiBearerAuth()
@UseGuards(JwtAuthGuard)
@Controller("admin/productos")
export class AdminProductoController {
  constructor(
    private readonly productoService: ProductoService,
    private readonly imagenService: ImagenService,
  ) {}

  @Post()
  @HttpCode(HttpStatus.CREATED)
  @UseInterceptors(
    FilesInterceptor("imagenes", 10, {
      storage: memoryStorage(),
      limits: { fileSize: 5 * 1024 * 1024 },
    }),
  )
  @ApiConsumes("multipart/form-data")
  @ApiOperation({ summary: "Publicar un nuevo producto (admin)" })
  @ApiBody({
    schema: {
      type: "object",
      required: ["nombre", "talle", "precioCentavos", "imagenes"],
      properties: {
        nombre: { type: "string", minLength: 3, example: "Remera Algodón" },
        descripcion: { type: "string", example: "Remera de algodón 100%" },
        talle: { type: "string", example: "M" },
        precioCentavos: { type: "number", example: 15000 },
        imagenes: {
          type: "array",
          items: { type: "string", format: "binary" },
        },
      },
    },
  })
  @ApiResponse({ status: 201, description: "Producto publicado exitosamente" })
  @ApiResponse({ status: 400, description: "Datos inválidos o sin imágenes" })
  @ApiResponse({ status: 401, description: "No autenticado" })
  async publicar(
    @Body() dto: PublicarProductoDto,
    @UploadedFiles() archivos: Express.Multer.File[],
  ) {
    const archivosValidos = archivos ?? [];

    if (archivosValidos.length === 0) {
      throw new BadRequestException("Seleccione al menos una imagen valida");
    }

    const urlsImagenes = await this.imagenService.guardar(archivosValidos);
    const producto = await this.productoService.publicar(dto, urlsImagenes);

    return {
      mensaje: "Producto publicado exitosamente",
      producto,
    };
  }

  @Get()
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: "Listar todos los productos (admin)" })
  @ApiResponse({ status: 200, description: "Lista de productos" })
  @ApiResponse({ status: 401, description: "No autenticado" })
  async listar() {
    return this.productoService.listar();
  }

  @Get(":id")
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: "Obtener un producto por ID (admin)" })
  @ApiResponse({ status: 200, description: "Producto encontrado" })
  @ApiResponse({ status: 401, description: "No autenticado" })
  @ApiResponse({ status: 404, description: "Producto no encontrado" })
  async obtenerUno(@Param("id") id: string) {
    return this.productoService.obtenerUno(id);
  }

  @Post("draft")
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ summary: "Crear producto en borrador (admin)" })
  @ApiResponse({ status: 201, description: "Producto borrador creado" })
  @ApiResponse({ status: 401, description: "No autenticado" })
  async crearDraft(@Body() dto: CreateProductoDto) {
    const producto = await this.productoService.crear(dto);
    return { mensaje: "Producto borrador creado exitosamente", producto };
  }

  @Patch(":id")
  @HttpCode(HttpStatus.OK)
  @UseInterceptors(
    FilesInterceptor("imagenes", 10, {
      storage: memoryStorage(),
      limits: { fileSize: 5 * 1024 * 1024 },
    }),
  )
  @ApiConsumes("multipart/form-data")
  @ApiOperation({ summary: "Actualizar un producto existente (admin)" })
  @ApiBody({
    schema: {
      type: "object",
      properties: {
        nombre: { type: "string", minLength: 3, example: "Remera Algodón" },
        descripcion: { type: "string", example: "Remera de algodón 100%" },
        talle: { type: "string", example: "M" },
        precioCentavos: { type: "number", example: 15000 },
        stock: { type: "number", example: 10 },
        imagenes: {
          type: "array",
          items: { type: "string", format: "binary" },
        },
      },
    },
  })
  @ApiResponse({
    status: 200,
    description: "Producto actualizado exitosamente",
  })
  @ApiResponse({
    status: 400,
    description: "Datos inválidos o imagen inválida",
  })
  @ApiResponse({ status: 401, description: "No autenticado" })
  @ApiResponse({ status: 404, description: "Producto no encontrado" })
  async actualizar(
    @Param("id") id: string,
    @Body() dto: ActualizarProductoDto,
    @UploadedFiles() archivos: Express.Multer.File[],
  ) {
    const archivosValidos = archivos ?? [];

    let urlsNuevasImagenes: string[] | undefined;
    let urlsImagenesAntiguas: string[] | undefined;

    if (archivosValidos.length > 0) {
      // Optimizamos para no llamar a obtenerUno() que construye una respuesta completa,
      // sino solo necesitamos las URLs de las imágenes actuales.
      const productoActual = await this.productoService.obtenerUno(id);
      urlsImagenesAntiguas = productoActual.imagenes;
      urlsNuevasImagenes = await this.imagenService.guardar(archivosValidos);
    }

    const producto = await this.productoService.actualizarAdmin(
      id,
      dto,
      urlsNuevasImagenes,
    );

    if (urlsImagenesAntiguas && urlsImagenesAntiguas.length > 0) {
      await this.imagenService.eliminar(urlsImagenesAntiguas);
    }

    return {
      mensaje: "Producto actualizado exitosamente",
      producto,
    };
  }

  @Delete(":id")
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: "Eliminar un producto (admin)" })
  @ApiResponse({ status: 200, description: "Producto eliminado" })
  @ApiResponse({ status: 401, description: "No autenticado" })
  @ApiResponse({ status: 404, description: "Producto no encontrado" })
  async eliminar(@Param("id") id: string) {
    await this.productoService.eliminar(id);
    return { mensaje: "Producto eliminado exitosamente" };
  }

  @Put(":id/desactivar")
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: "Desactivar un producto (admin)" })
  @ApiResponse({
    status: 200,
    description: "Producto desactivado exitosamente",
  })
  @ApiResponse({ status: 400, description: "El producto ya está desactivado" })
  @ApiResponse({ status: 401, description: "No autenticado" })
  @ApiResponse({ status: 404, description: "Producto no encontrado" })
  async desactivar(@Param("id") id: string) {
    const producto = await this.productoService.desactivar(id);
    return {
      mensaje: "Producto desactivado exitosamente",
      producto,
    };
  }
}

File: backend/src/main/java/com/sistemapedidos/modules/producto/AdminProductoController.java
package com.sistemapedidos.modules.producto;

import com.sistemapedidos.common.error.BadRequestException;
import com.sistemapedidos.modules.producto.dto.ActualizarProductoDto;
import com.sistemapedidos.modules.producto.dto.CreateProductoDto;
import com.sistemapedidos.modules.producto.dto.PaginadoProductos;
import com.sistemapedidos.modules.producto.dto.ProductoResponse;
import com.sistemapedidos.modules.producto.dto.PublicarProductoDto;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/admin/productos")
@RequiredArgsConstructor
public class AdminProductoController {

    private final ProductoService productoService;
    private final ImagenService imagenService;

    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    @ResponseStatus(HttpStatus.CREATED)
    public Map<String, Object> publicar(@Valid @ModelAttribute PublicarProductoDto dto,
                                        @RequestParam(value = "imagenes", required = false) List<MultipartFile> archivos) {
        List<MultipartFile> archivosValidos = archivos != null ? archivos : List.of();

        if (archivosValidos.isEmpty()) {
            throw new BadRequestException("Seleccione al menos una imagen valida");
        }

        List<String> urlsImagenes = imagenService.guardar(archivosValidos);
        ProductoResponse producto = productoService.publicar(dto, urlsImagenes);

        return Map.of("mensaje", "Producto publicado exitosamente", "producto", producto);
    }

    @GetMapping
    public PaginadoProductos listar() {
        return productoService.listar(null, null, null, 1, 20);
    }

    @GetMapping("/{id}")
    public ProductoResponse obtenerUno(@PathVariable String id) {
        return productoService.obtenerUno(id);
    }

    @PostMapping("/draft")
    @ResponseStatus(HttpStatus.CREATED)
    public Map<String, Object> crearDraft(@Valid @RequestBody CreateProductoDto dto) {
        ProductoResponse producto = productoService.crear(dto);
        return Map.of("mensaje", "Producto borrador creado exitosamente", "producto", producto);
    }

    @PatchMapping(value = "/{id}", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public Map<String, Object> actualizar(@PathVariable String id,
                                          @Valid @ModelAttribute ActualizarProductoDto dto,
                                          @RequestParam(value = "imagenes", required = false) List<MultipartFile> archivos) {
        List<MultipartFile> archivosValidos = archivos != null ? archivos : List.of();

        List<String> urlsNuevasImagenes = null;
        List<String> urlsImagenesAntiguas = null;

        if (!archivosValidos.isEmpty()) {
            ProductoResponse productoActual = productoService.obtenerUno(id);
            urlsImagenesAntiguas = productoActual.imagenes();
            urlsNuevasImagenes = imagenService.guardar(archivosValidos);
        }

        ProductoResponse producto = productoService.actualizarAdmin(id, dto, urlsNuevasImagenes);

        if (urlsImagenesAntiguas != null && !urlsImagenesAntiguas.isEmpty()) {
            imagenService.eliminar(urlsImagenesAntiguas);
        }

        return Map.of("mensaje", "Producto actualizado exitosamente", "producto", producto);
    }

    @DeleteMapping("/{id}")
    public Map<String, String> eliminar(@PathVariable String id) {
        productoService.eliminar(id);
        return Map.of("mensaje", "Producto eliminado exitosamente");
    }

    @PutMapping("/{id}/desactivar")
    public Map<String, Object> desactivar(@PathVariable String id) {
        ProductoResponse producto = productoService.desactivar(id);
        return Map.of("mensaje", "Producto desactivado exitosamente", "producto", producto);
    }
}
